File: web-face/src/app/core/services/search.service.ts
import { inject, Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

import { environment } from '../../../environments/environment';
import { Page } from '../models/page.model';
import { SearchHit, SearchResultType } from '../models/search.model';
import { buildHttpParams } from './http-params.util';

export interface SearchParams {
  /** Free-text match against name/tags. */
  q?: string;
  type?: SearchResultType;
  category?: string;
  tag?: string;
  page?: number;
  size?: number;
}

/** Cross-entity search over Location and Device, backing GET /api/v1/search. */
@Injectable({ providedIn: 'root' })
export class SearchService {
  private readonly http = inject(HttpClient);

  search(params: SearchParams): Observable<Page<SearchHit>> {
    return this.http.get<Page<SearchHit>>(`${environment.apiUrl}/search`, {
      params: buildHttpParams(params),
    });
  }
}

File: core/src/main/java/com/csl/lasform/service/SearchService.java
package com.csl.lasform.service;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import com.csl.lasform.model.search.SearchHit;
import com.csl.lasform.model.search.SearchResultType;

/** Cross-entity search over Location and Device. */
public interface SearchService {

    /**
     * @param type     restricts the search to one entity kind; {@code null} searches both
     * @param category matches entities whose {@code categoryIds} contains this value, if given
     * @param tag      matches entities whose {@code tags} contains this value, if given
     * @param q        case-insensitive free-text match against {@code name} or any {@code tags} entry, if given
     */
    Page<SearchHit> search(SearchResultType type, String category, String tag, String q, Pageable pageable);
}
